### ISSUE TITLE
PEP-695 type parameters are incorrectly rebound to methods in nested classes

### ISSUE THREAD

**Issue by AlexWaygood** [astral-sh/ty#4559]
### Summary

A method of a nested class incorrectly treats an enclosing class’s PEP 695 type parameter as belonging to the method.

```py
from typing import reveal_type

class Outer[T]:
    class Inner:
        def method(self, value: T) -> T:
            return value

reveal_type(Outer.Inner.method)
```

ty reveals:

```text
def method(self, value: T@method) -> T@method
```

Both occurrences of `T` should refer to `T@Outer`. `method` does not declare any type parameters, and PEP 695 parameters have explicit lexical scopes: they cannot implicitly acquire a new binding when
referenced in a nested method.

[PEP 695’s scoping rules](https://peps.python.org/pep-0695/#type-parameter-scopes) explicitly include an example where a method of a nested class references the outer class’s type parameter. The rules differ
from legacy `TypeVar` declarations, which can introduce independent bindings when reused.

Pyright preserves the expected binding:

```text
(self: Inner, value: T@Outer) -> T@Outer
```

The likely cause is in [`bind_typevar` and `find_typevar_binding`](https://github.com/astral-sh/ruff/blob/3f79413aadcdf1dcfa9bd7c4e82989669f4b518c/crates/ty_python_semantic/src/types/generics.rs#L90-L250). The
lookup hides outer class bindings after crossing an inner class boundary, then falls back to binding the variable to the current function. This behavior also applies to PEP 695 parameters, despite their
declaration fixing which generic context owns them.

Reproduced at `astral-sh/ruff@3f79413aadcdf1dcfa9bd7c4e82989669f4b518c` with `--python-version 3.12`.
